packages/kbn-config-schema/src/types/object_type.test.ts

```diff
@@ -490,3 +490,76 @@ describe('#extends', () => {
     expect(extended.validate(undefined)).toEqual({ initial: 'bar', added: 42 });
   });
 });
+
+test('returns schema structure', () => {
+  // This test covers different schema types that may or may not be nested
+  const objSchema = schema.object({
+    any: schema.any(),
+    array: schema.arrayOf(schema.string()),
+    boolean: schema.boolean(),
+    buffer: schema.buffer(),
+    byteSize: schema.byteSize(),
+    conditional: schema.conditional(
+      schema.contextRef('context_value_1'),
+      schema.contextRef('context_value_2'),
+      schema.string(),
+      schema.string()
+    ),
+    duration: schema.duration(),
+    ip: schema.ip(),
+    literal: schema.literal('foo'),
+    map: schema.mapOf(schema.string(), schema.string()),
+    maybe: schema.maybe(schema.string()),
+    never: schema.never(),
+    nullable: schema.nullable(schema.string()),
+    number: schema.number(),
+    record: schema.recordOf(schema.string(), schema.string()),
+    stream: schema.stream(),
+    string: schema.string(),
+    union: schema.oneOf([schema.string()]),
+    uri: schema.uri(),
+  });
+  const type = objSchema.extends({
+    nested: objSchema,
+  });
+  expect(type.getSchemaStructure()).toEqual([
+    { path: ['any'], type: 'any' },
+    { path: ['array'], type: 'array' },
+    { path: ['boolean'], type: 'boolean' },
+    { path: ['buffer'], type: 'binary' },
+    { path: ['byteSize'], type: 'bytes' },
+    { path: ['conditional'], type: 'any' },
+    { path: ['duration'], type: 'duration' },
+    { path: ['ip'], type: 'string' },
+    { path: ['literal'], type: 'any' },
+    { path: ['map'], type: 'map' },
+    { path: ['maybe'], type: 'any' },
+    { path: ['never'], type: 'any' },
+    { path: ['nullable'], type: 'alternatives' },
+    { path: ['number'], type: 'number' },
+    { path: ['record'], type: 'record' },
+    { path: ['stream'], type: 'stream' },
+    { path: ['string'], type: 'any' },
```

Comment:
I didn't understand why `uri` was property recognized as `string` where `string` wasn't so I took a look at our `string` type implementation. 

Because of a limitation on joi's `allow`, we were forced to use a custom schema. But we can actually force the type of the custom schema to be recognized as `string`:

```diff
diff --git a/packages/kbn-config-schema/src/types/string_type.ts b/packages/kbn-config-schema/src/types/string_type.ts
--- a/packages/kbn-config-schema/src/types/string_type.ts	(revision 62e06600f790fc487365284e655ca0fab9a59272)
+++ b/packages/kbn-config-schema/src/types/string_type.ts	(date 1649405961321)
@@ -53,6 +53,7 @@
       );
     }
 
+    schema.type = 'string';
     super(schema, options);
   }

```

I opened a PR just to make sure this change would not be breaking anything by side effect: https://github.com/elastic/kibana/pull/129841

If the PR comes up green, would you mind integrating the change in your PR?

Reply:
Nice! Sure thing